Add spec for staking

foriequal0 opened the pull request.

sgkim126 commented on spec/Staking.md:
````diff
@@ -0,0 +1,192 @@
+  }
+}
+```
+In a such state, an account "0x23..45" has a weight of 500 (balance 450 + delegation 50), and "0xAB..CD" hass a weight of 1000 (balance 690 + delegation 200 + delegation 110). If a block has total fees of 110, each account will share following amount of fees.
````
Please apply the change of https://github.com/CodeChain-io/codechain/pull/1426

ScarletBlue commented on spec/Staking.md:
```diff
@@ -0,0 +1,340 @@
+
+# Fee distribution
+
+You pay fees to make a transaction, which is greater than a specified minimum fee for the transaction type. Transaction fees within a block are collected and CCS holders and the block author share this.
```
You pay fees to make a transaction, which is greater than the specified minimum fee for the transaction type. Transaction fees that are within a block are collected, and CCS holders and the block author share this.

ScarletBlue commented on spec/Staking.md:
```diff
@@ -0,0 +1,340 @@
+
+You pay fees to make a transaction, which is greater than a specified minimum fee for the transaction type. Transaction fees within a block are collected and CCS holders and the block author share this.
+
+CCS Holders share the sum of the minimum fees for the transaction. They get fees in proportion to the sum of the balance of CCS plus the total amounts of CCS that are delegated to someone.
```
CCS Holders share the sum of the minimum fees for the transaction. They get fees in proportion to the sum of the balance of CCS plus the total amount of CCS that are delegated to someone.

ScarletBlue commented on spec/Staking.md:
````diff
@@ -0,0 +1,340 @@
+}
+```
+
+In a such state, an account "0x23..45" has a weight of 500 (balance 450 + delegation 50), and "0xAB..CD" has a weight of 1000 (balance 690 + delegation 200 + delegation 110, pending revocation will not affect to it). If a block has total fees of 110, and if the sum of the minimum fees for it is 35, each account will share the following amount of fees.
````
In a such state, an account "0x23..45" has a weight of 500 (balance 450 + delegation 50), and "0xAB..CD" has a weight of 1000 (balance 690 + delegation 200 + delegation 110, pending revocation will not affect it). If a block has total fees of 110, and if the sum of the minimum fees for it is 35, each account will share the following amount of fees.

ScarletBlue commented on spec/Staking.md:
```diff
@@ -0,0 +1,340 @@
+
+# Revocation
+
+`BeginRevoke` will queue a pending revocation instead of not revoking a delegation immediately. The revocation will be delayed to a certain amount of time (`REVOKE_PENDDING_INTERVAL`) to prevent abuse. It will be processed when a block whose timestamp is greater or equal than `endTime` is created. The balance of `delegator` will be increased by `quantity` on the revocation, and the pending revoca ...
```
`BeginRevoke` will queue a pending revocation instead of revoking a delegation immediately. The revocation will be delayed by a certain amount of time (`REVOKE_PENDDING_INTERVAL`) to prevent abuse. It will be processed when a block whose timestamp is greater than or equal to the time when `endTime` is created. The balance of `delegator` will be increased by `quantity` upon revocation, and the pending revocation will be removed from the revocation queue. The queue and delegations that become empty should be removed from the state.

foriequal0 commented:
I changed some terms to clarify concepts and fixed the descriptions of behaviors to match with current implementations. I'll amend last commit after reviews.

ScarletBlue commented on spec/Staking.md:
```diff
@@ -0,0 +1,368 @@
+
+    A `receiver` will be inserted to the list of CCS holders, and its amount of undelegated CCS will be increased by `quantity`.
+    The transaction sender's amount of undelegated CCS will be decreased by `quantity`.
+    A sender cannot transfer more than what its amount of undelegated CCS.
```
A sender cannot transfer more than the amount of undelegated CCS it has.

ScarletBlue commented on spec/Staking.md:
```diff
@@ -0,0 +1,368 @@
+
+    A `delegatee` must be one of the current validators.
+    The amount of undelegated CCS of the sender will be decreased by `quantity`. However, instead of increasing the amount of undelegated CCS of the `delegatee`, [`delegatee`, `quantity`] will be inserted to the transaction sender's delegation list.
+    A sender cannot delegate more than what its amount of undelegated CCS.
```
A sender cannot delegate more than what its amount of undelegated CCS is.

ScarletBlue commented on spec/Staking.md:
```diff
@@ -0,0 +1,368 @@
+
+# Fee distribution
+
+You pay fees to make a transaction, which is greater than the specified minimum fee for the transaction type.
```
You pay fees to make a transaction. Fees should be greater than the specified minimum fee for the transaction type.

ScarletBlue commented on spec/Staking.md:
```diff
@@ -0,0 +1,368 @@
+Transaction fees that are within a block are collected, and CCS holders and the block author share this.
+
+CCS Holders share the sum of the minimum fees for the transaction.
+They get fees in proportion to the CCS balance they have which is the sum of the amount of undelegated CCS plus the total amount of CCS that are delegated to someone.
```
They get fees in proportion to the CCS balance they have, which is the sum of the amount of undelegated CCS plus the total amount of CCS that are delegated to someone.

ScarletBlue commented on spec/Staking.md:
````diff
@@ -0,0 +1,368 @@
+}
+```
+
+In a such state, an account "0x23..45" has the balance of 500 (undelegated 450 + delegated 50), and "0xAB..CD" has the balance of 1000 (undelegated 690 + delegated 200 + delegated 110, pending revocation will not affect it).
````
In a such state, an account "0x23..45" has a balance of 500 (undelegated 450 + delegated 50), and "0xAB..CD" has a balance of 1000 (undelegated 690 + delegated 200 + delegated 110, pending revocation will not affect it).

ScarletBlue commented on spec/Staking.md:
````diff
@@ -0,0 +1,368 @@
+```
+
+In a such state, an account "0x23..45" has the balance of 500 (undelegated 450 + delegated 50), and "0xAB..CD" has the balance of 1000 (undelegated 690 + delegated 200 + delegated 110, pending revocation will not affect it).
+If a block has total fees of 110, and the sum of the minimum fees for it is 35, each account will share the following amount of fees.
````
If a block has total fees of 110, and the sum of the minimum fees for it is 35, each account will share the following amount of fees:

ScarletBlue commented on spec/Staking.md:
```diff
@@ -0,0 +1,368 @@
+The revocation will be delayed by a certain amount of time (`REVOKE_PENDING_DURATION`) to prevent abuse.
+It will be processed when a block whose timestamp is greater than or equal to the time when `endTime` is created.
+The amount of undelegated CCS of the `delegator` will be increased by `quantity` upon revocation, and the amount of delegated CCS that the `delegator` has delegated to the `delegatee` will be decreased by `quantity`, and the pending revocation will be removed from the revocation queue.
+The queue and delegations that become empty should be removed from the state.
```
Queues and delegations that become empty should be removed from the state.
